app/scripts/lib/SnapsNameProvider.ts

```diff
@@ -0,0 +1,158 @@
+import {
+  NameProvider,
+  NameProviderMetadata,
+  NameProviderRequest,
+  NameProviderResult,
+  NameProviderSourceResult,
+  NameType,
+} from '@metamask/name-controller';
+import { GetPermissionControllerState } from '@metamask/permission-controller';
+import {
+  OnNameLookupArgs,
+  HandlerType,
+  OnNameLookupResponse,
+  TruncatedSnap,
+} from '@metamask/snaps-utils';
+import log from 'loglevel';
+import {
+  GetAllSnaps,
+  GetSnap,
+  HandleSnapRequest,
+} from '@metamask/snaps-controllers';
+import { RestrictedControllerMessenger } from '@metamask/base-controller';
+
+type AllowedActions =
+  | GetAllSnaps
+  | GetSnap
+  | HandleSnapRequest
+  | GetPermissionControllerState;
+
+export type SnapsNameProviderMessenger = RestrictedControllerMessenger<
+  'SnapsNameProvider',
+  AllowedActions,
+  never,
+  AllowedActions['type'],
+  never
+>;
+
+export class SnapsNameProvider implements NameProvider {
+  #messenger: SnapsNameProviderMessenger;
+
+  constructor({ messenger }: { messenger: SnapsNameProviderMessenger }) {
+    this.#messenger = messenger;
+  }
+
+  getMetadata(): NameProviderMetadata {
+    const snaps = this.#getNameLookupSnaps();
+
+    const sourceIds = {
+      [NameType.ETHEREUM_ADDRESS]: snaps.map((snap) => snap.id),
+    };
+
+    const sourceLabels = snaps.reduce(
+      (acc: NameProviderMetadata['sourceLabels'], snap) => {
+        const snapDetails = this.#messenger.call('SnapController:get', snap.id);
+        const snapName = snapDetails?.manifest.proposedName;
```

Comment:
May want to consider using `getSnapName`

Reply:
Good point, will add a task to our backlog to move `getSnapName` into shared and use it here.